staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval_test.go

```diff
@@ -0,0 +1,486 @@
+	postOkCheck          func() bool
+	received, expected   int
+	event, expectedEvent *watchCacheEvent
+	nilEventSanityCheck  func() bool
+}
+
+func (ev eventVerifier) verify() error {
+	if !ev.ok {
+		if ev.received != ev.expected {
+			return errors.New("unexpected bool value indicating empty buffer")
+		}
+		return nil
+	}
+
+	if ev.postOkCheck != nil {
+		if ev.postOkCheck() {
+			return errors.New("unexpected bool value indicating non-empty buffer")
+		}
+	}
+
+	if ev.event == nil {
+		if ev.nilEventSanityCheck != nil {
+			if pass := ev.nilEventSanityCheck(); !pass {
+				return fmt.Errorf("unexpected nil event, expected: %v", *ev.expectedEvent)
+			}
+			return nil
+		}
+		return fmt.Errorf("unexpected nil event, expected: %v", *ev.expectedEvent)
+	}
+
+	if !reflect.DeepEqual(ev.event, ev.expectedEvent) {
+		return fmt.Errorf("expected %v, got %v", *ev.event, *ev.expectedEvent)
+	}
+
+	return nil
+}
+
+func verifyNoEvent(ok bool, event *watchCacheEvent) error {
+	if ok {
+		return errors.New("unexpected bool value indicating buffer is not empty")
+	}
+	if event != nil {
+		return fmt.Errorf("unexpected event received, expected: nil, got %v", *event)
+	}
+
+	return nil
+}
+
+func TestIntervalBufferIsFull(t *testing.T) {
+	cases := []struct {
+		endIndex int
+		expected bool
+	}{
+		{endIndex: bufferSize - 1, expected: false},
+		{endIndex: bufferSize, expected: true},
+		{endIndex: bufferSize + 1, expected: true},
+	}
+
+	for _, c := range cases {
+		wcib := &watchCacheIntervalBuffer{endIndex: c.endIndex}
+		actual := wcib.isFull()
+		if actual != c.expected {
+			t.Errorf("expected %v, got %v", c.expected, actual)
+		}
+	}
+}
+
+func TestIntervalBufferIsEmpty(t *testing.T) {
+	cases := []struct {
+		startIndex int
+		endIndex   int
+		expected   bool
+	}{
+		{startIndex: 0, endIndex: 10, expected: false},
+		{startIndex: 5, endIndex: 20, expected: false},
+		{startIndex: 50, endIndex: 50, expected: true},
+	}
+
+	for _, c := range cases {
+		wcib := &watchCacheIntervalBuffer{
+			startIndex: c.startIndex,
+			endIndex:   c.endIndex,
+		}
+		actual := wcib.isEmpty()
+		if actual != c.expected {
+			t.Errorf("expected %v, got %v", c.expected, actual)
+		}
+	}
+}
+
+func TestIntervalBufferNext(t *testing.T) {
+	cases := []struct {
+		name   string
+		events []*watchCacheEvent
+	}{
+		{
+			name: "buffer has elements",
+			events: []*watchCacheEvent{
+				{Type: watch.Added, Object: makeTestPod("pod1", 1)},
+				{Type: watch.Added, Object: makeTestPod("pod2", 2)},
+				{Type: watch.Modified, Object: makeTestPod("pod3", 3)},
+			},
+		},
+		{
+			name:   "buffer is empty",
+			events: []*watchCacheEvent{},
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			wcib := bufferFromEvents(c.events)
+
+			for i := 0; i < len(c.events); i++ {
+				event, ok := wcib.next()
+				verifier := eventVerifier{
+					ok:            ok,
+					postOkCheck:   func() bool { return i == len(c.events) },
+					received:      i,
```

Comment:
explaining comment : I don't think we need to check received/expected, because we track it exactly by for loops (we need exactly how many items we process and we verify at the end verifyNoEvent).
So those doesn't provide any value.